cmd/zfs/zfs_main.c

```diff
@@ -904,7 +904,7 @@ zfs_do_create(int argc, char **argv)
 			if (nvlist_add_uint64(props,
 			    zfs_prop_to_name(ZFS_PROP_VOLSIZE), intval) != 0)
 				nomem();
-			volsize = intval;
+			volsize = P2ROUNDUP(intval, align);
```

Comment:
This macro requires you to pass in the requested power-of-two alignment as the `align` parameter above (in this case the block size).  Which is going to be a bit of a problem since at this point in we're still parsing arguments so we can't know if they're provided a different block size.  Plus it looks like setting the `volsize` value here isn't going to be enough.  The volume size is passed to the kernel via the `ZFS_PROP_VOLSIZE` property in the nvlist a few lines up, so that will need to be updated as well.

The best way to go about handling this would be to move the needed logic down after the argument parsing where we have access to the volume size and block size.  It's probably worth adding a small helper function to do what's needed, that `zvol_volsize_to_reservation()` provides a nice example.